lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
+
+    // TODO: Set response type here
+    acquireTokenAuthority.ResolveEndpointsAsync().then(() => {
+      // On Fulfillment
+      if (Utils.compareObjects(userObject, this.getUser())) {
+        if (scopes.indexOf(this.clientId) > -1) {
+          authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), this._state);
+        } else {
+          authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.token, this.getRedirectUri(), this._state);
+        }
+      } else {
+        authenticationRequest = new AuthenticationRequestParameters(acquireTokenAuthority, this.clientId, scopes, ResponseTypes.id_token_token, this.getRedirectUri(), this._state);
+      }
+
+      // Cache nonce
+      this._cacheStorage.setItem(Constants.nonceIdToken, authenticationRequest.nonce, this.storeAuthStateInCookie);
+
+      // Cache acquireTokenUserKey
+      var acquireTokenUserKey;
+      if (userObject) {
+        acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter + userObject.userIdentifier + Constants.resourceDelimeter + authenticationRequest.state;
+      } else {
+        acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter  + Constants.no_user + Constants.resourceDelimeter + authenticationRequest.state;
+      }
+      this._cacheStorage.setItem(acquireTokenUserKey, JSON.stringify(userObject));
+
+      // Cache authorityKey
+      const authorityKey = Constants.authority + Constants.resourceDelimeter + authenticationRequest.state;
+      this._cacheStorage.setItem(authorityKey, acquireTokenAuthority.CanonicalAuthority, this.storeAuthStateInCookie);
+
+      // Set extraQueryParameters to be sent to the server
+      if (extraQueryParameters) {
+        authenticationRequest.extraQueryParameters = extraQueryParameters;
+      }
+
+      // Construct urlNavigate
+      let urlNavigate = authenticationRequest.createNavigateUrl(scopes)   + Constants.response_mode_fragment;
+      urlNavigate = this.addHintParameters(urlNavigate, userObject);
+
+      // set state in cache and redirect to urlNavigate
+      if (urlNavigate) {
+        this._cacheStorage.setItem(Constants.stateAcquireToken, authenticationRequest.state, this.storeAuthStateInCookie);
+        window.location.replace(urlNavigate);
+      }
+    });
+  }
+
+  /** 
+   * Checks if the redirect response is received from the STS. In case of redirect, the url fragment has either id_token, access_token or error.
+   * @param {string} hash - Hash passed from redirect page.
+   * @returns {Boolean} - true if response contains id_token, access_token or error, false otherwise.
+   * @hidden
+   */
+  // TODO - rename this, the name is confusing
+  isCallback(hash: string): boolean {
```

Comment:
also rename the parameter `hash` is passed in but then use a method called gethash from hash?

Reply:
+1